### PR TITLE
Refactoring in builder and fix some nested status fields issue

### LINKED ISSUE #27: Some nested status fields might be missing
### What problem are you facing?

We use Terraform schema to generate the type of every field. While Terraform has one big untyped tree of fields to hold all the information, that's not the case in Crossplane. In line with Kubernetes API conventions, we separate fields into `spec` and `status` fields, `status` ones being fields that user cannot ever configure.

In Terraform schema, fields report whether they are `Computed` and `Optional`. The fields that are both `Computed` and `Optional` are the ones that have server-side defaults and user-configurable; the ones we late-initialize. So we [put those under spec](https://github.com/crossplane-contrib/terrajet/blob/df8f6cf/pkg/types/builder.go#L87), but fields that are `Computed` and `not Optional` [go to `status`](https://github.com/crossplane-contrib/terrajet/blob/df8f6cf/pkg/types/builder.go#L168).

The problem is that [there can be](https://github.com/crossplane-contrib/terrajet/blob/df8f6cf/pkg/types/builder.go#L155) some types that are `Computed` and `not Optional` but also has nested fields that can go under `spec` and vice versa. Right now, when we compute the type for a field, we return a parameters type as well as an observation type, then decide which one to use depending on the properties of the field. If the field goes to spec but there are nested fields that could go to status, they are eliminated. Also, if a field goes to status, all nested fields go to status as well and the parameter types are ignored.

When I did an analysis of how prevalent this problem is in, for example, AWS, I see the following output:
```
there are 1 fields that could be observation type of .SyntheticsCanary.VpcConfig but ignored since the field is a parameters field
there are 2 fields that could be observation type of .AmplifyDomainAssociation.SubDomain but ignored since the field is a parameters field
there are 1 fields that could be observation type of .StoragegatewayGateway.SmbActiveDirectorySettings but ignored since the field is a parameters field
there are 1 fields that could be observation type of .LexBotAlias.ConversationLogs.LogSettings but ignored since the field is a parameters field
there are 1 fields that could be observation type of .Route53ResolverEndpoint.IpAddress but ignored since the field is a parameters field
there are 1 fields that could be observation type of .LambdaFunction.VpcConfig but ignored since the field is a parameters field
there are 1 fields that could be observation type of .MwaaEnvironment.LoggingConfiguration.DagProcessingLogs but ignored since the field is a parameters field
there are 1 fields that could be observation type of .MwaaEnvironment.LoggingConfiguration.SchedulerLogs but ignored since the field is a parameters field
there are 1 fields that could be observation type of .MwaaEnvironment.LoggingConfiguration.TaskLogs but ignored since the field is a parameters field
there are 1 fields that could be observation type of .MwaaEnvironment.LoggingConfiguration.WebserverLogs but ignored since the field is a parameters field
there are 1 fields that could be observation type of .MwaaEnvironment.LoggingConfiguration.WorkerLogs but ignored since the field is a parameters field
there are 4 fields that could be parameters type of .ElasticBeanstalkEnvironment.AllSettings but ignored since the field is an observation field
there are 2 fields that could be observation type of .EksCluster.VpcConfig but ignored since the field is a parameters field
there are 2 fields that could be observation type of .DirectoryServiceDirectory.ConnectSettings but ignored since the field is a parameters field
there are 1 fields that could be observation type of .DirectoryServiceDirectory.VpcSettings but ignored since the field is a parameters field
there are 3 fields that could be observation type of .EmrCluster.CoreInstanceFleet but ignored since the field is a parameters field
there are 1 fields that could be observation type of .EmrCluster.CoreInstanceGroup but ignored since the field is a parameters field
there are 3 fields that could be observation type of .EmrCluster.MasterInstanceFleet but ignored since the field is a parameters field
there are 1 fields that could be observation type of .EmrCluster.MasterInstanceGroup but ignored since the field is a parameters field
there are 2 fields that could be observation type of .Kinesisanalyticsv2Application.ApplicationConfiguration.SqlApplicationConfiguration.Input but ignored since the field is a parameters field
there are 1 fields that could be observation type of .Kinesisanalyticsv2Application.ApplicationConfiguration.SqlApplicationConfiguration.Output but ignored since the field is a parameters field
there are 1 fields that could be observation type of .Kinesisanalyticsv2Application.ApplicationConfiguration.SqlApplicationConfiguration.ReferenceDataSource but ignored since the field is a parameters field
there are 2 fields that could be observation type of .Kinesisanalyticsv2Application.ApplicationConfiguration.VpcConfiguration but ignored since the field is a parameters field
there are 1 fields that could be observation type of .Kinesisanalyticsv2Application.CloudwatchLoggingOptions but ignored since the field is a parameters field
there are 3 fields that could be observation type of .SecretsmanagerSecret.Replica but ignored since the field is a parameters field
there are 4 fields that could be parameters type of .SsmDocument.Parameter but ignored since the field is an observation field
there are 1 fields that could be observation type of .Alb.SubnetMapping but ignored since the field is a parameters field
there are 1 fields that could be observation type of .SpotInstanceRequest.EbsBlockDevice but ignored since the field is a parameters field
there are 2 fields that could be observation type of .SpotInstanceRequest.RootBlockDevice but ignored since the field is a parameters field
there are 1 fields that could be observation type of .Lb.SubnetMapping but ignored since the field is a parameters field
there are 2 fields that could be observation type of .ElasticsearchDomain.VpcOptions but ignored since the field is a parameters field
there are 2 fields that could be observation type of .Apigatewayv2DomainName.DomainNameConfiguration but ignored since the field is a parameters field
there are 1 fields that could be observation type of .NetworkInterface.Attachment but ignored since the field is a parameters field
there are 1 fields that could be observation type of .Instance.EbsBlockDevice but ignored since the field is a parameters field
there are 2 fields that could be observation type of .Instance.RootBlockDevice but ignored since the field is a parameters field
there are 1 fields that could be observation type of .CodestarnotificationsNotificationRule.Target but ignored since the field is a parameters field
there are 2 fields that could be observation type of .CodeartifactRepository.ExternalConnections but ignored since the field is a parameters field
there are 8 fields that could be observation type of .AmiCopy.EbsBlockDevice but ignored since the field is a parameters field
there are 2 fields that could be observation type of .AmiCopy.EphemeralBlockDevice but ignored since the field is a parameters field
there are 8 fields that could be observation type of .AmiFromInstance.EbsBlockDevice but ignored since the field is a parameters field
there are 2 fields that could be observation type of .AmiFromInstance.EphemeralBlockDevice but ignored since the field is a parameters field
there are 1 fields that could be observation type of .KinesisAnalyticsApplication.CloudwatchLoggingOptions but ignored since the field is a parameters field
there are 1 fields that could be observation type of .KinesisAnalyticsApplication.Inputs.Schema.RecordFormat but ignored since the field is a parameters field
there are 2 fields that could be observation type of .KinesisAnalyticsApplication.Inputs but ignored since the field is a parameters field
there are 1 fields that could be observation type of .KinesisAnalyticsApplication.Outputs but ignored since the field is a parameters field
there are 1 fields that could be observation type of .KinesisAnalyticsApplication.ReferenceDataSources.Schema.RecordFormat but ignored since the field is a parameters field
there are 1 fields that could be observation type of .KinesisAnalyticsApplication.ReferenceDataSources but ignored since the field is a parameters field
there are 1 fields that could be observation type of .KinesisFirehoseDeliveryStream.ElasticsearchConfiguration.VpcConfig but ignored since the field is a parameters field
there are 1 fields that could be observation type of .GlueCatalogTable.PartitionIndex but ignored since the field is a parameters field

Generated 770 resources!
```

There are thousands of fields generated for 770 resources, so problem is not that big, at least for AWS. We see that almost all ignored ones are the ones that exist under a parameters type but should've gone to observation. So, the problem does not cause much miss in the configuration ability of the users but there is information they'd not be able to see under status.

### How could Terrajet help solve your problem?

There can be several solutions. The most basic one is to have the full schema in both status and spec but that'd mean deviation from Crossplane Resource Model even if unused fields will stay empty and it's not really worth doing this for all resources to be correct for minority cases.

The best solution would be the one that would construct the earlier fields and types under status once we encounter an observational type under a parameter type (and vice versa). However, it's not done in the first iteration since the field path wasn't known by the schema builder. With https://github.com/crossplane-contrib/terrajet/pull/24 it will know the path.

So, we could possibly have the type map be a more complex tree with some utilities that will allow you to add a certain field to an arbitrary point and construct the path from the root to that leaf by building the necessary types & fields. Right now, it has the list of types for uniqueness purposes but the actual information is recursively saved in the Go type objects. So this solution would mean we'd deviate from the Go types package to store the information and that'd require some refactoring.

I'd propose that we can have a utility that will work on Go types library objects to manipulate the existing type trees to add arbitrary field to a field path. There would be a function that knows the root types and when you give it an arbitrary field  information, it'd  construct the path in the either of the two root structs (parameters and observation). There are some caveats about naming of those new fields and types but it's doable and lets us still be in the same type system instead of building our own.

### PR DESCRIPTION
**Pull request by sergenyalcin**
### Description of your changes

This PR contains refactoring in `builder.go`. We have a long function: `buildResource` This PR separates this function to new small functions. For being more modular, new structs were added.

Also,
Fixes #27 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

This change was tested by generating full-set of resources in jet-aws provider. This generation was successful. No difference was observed, except for the changes related to the fix of #27 

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by sergenyalcin** [comment]
Hi @muvaf @turkenh @ulucinar, this is a draft PR to refactor the builder function. Let's discuss the suggestions, possible improvements etc. Thanks!

**Review by muvaf — COMMENTED** [review]
Sorry for the late review on this big PR. It looks like a great start @sergenyalcin !

File: pkg/types/builder.go [hunk 1]
```diff
@@ -225,16 +167,131 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
-	obsType := types.NewNamed(obsName, types.NewStruct(obsFields, obsTags), nil)
+	obsType := types.NewNamed(obsName, types.NewStruct(r.obsFields, r.obsTags), nil)
 	g.genTypes = append(g.genTypes, obsType)
 
 	return paramType, obsType, nil
 }
 
-func (g *Builder) buildSchema(sch *schema.Schema, cfg *config.Resource, tfPath []string, xpPath []string, names []string) (types.Type, error) { // nolint:gocyclo
+func (g *Builder) getParameterAndObservationName(names []string) (*types.TypeName, *types.TypeName, error) {
+	paramTypeName, err := g.generateTypeName("Parameters", names...)
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: A new Namer struct was created

File: pkg/types/builder.go [hunk 2]
```diff
@@ -225,16 +167,131 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
 
-	obsType := types.NewNamed(obsName, types.NewStruct(obsFields, obsTags), nil)
+	obsType := types.NewNamed(obsName, types.NewStruct(r.obsFields, r.obsTags), nil)
 	g.genTypes = append(g.genTypes, obsType)
 
 	return paramType, obsType, nil
 }
 
-func (g *Builder) buildSchema(sch *schema.Schema, cfg *config.Resource, tfPath []string, xpPath []string, names []string) (types.Type, error) { // nolint:gocyclo
+func (g *Builder) getParameterAndObservationName(names []string) (*types.TypeName, *types.TypeName, error) {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Done!

File: pkg/types/builder.go [hunk 3]
```diff
@@ -225,16 +167,131 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
+
+	// We insert them to the package scope so that the type name calculations in
+	// recursive calls are checked against their upper level type's name as well.
+	g.Package.Scope().Insert(paramName)
+	g.Package.Scope().Insert(obsName)
+
+	return paramName, obsName, nil
+}
+
+func prepareField(res *schema.Resource, snakeFieldName string, tfPath, xpPath, names []string) (*field, error) {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Done!

File: pkg/types/builder.go [hunk 4]
```diff
@@ -225,16 +167,131 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
+
+	comment, err := comments.New(f.sch.Description)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot build comment for description: %s", f.sch.Description)
+	}
+	f.comment = comment
+	f.tfTag = fmt.Sprintf("%s,omitempty", f.fieldName.Snake)
+	f.jsonTag = fmt.Sprintf("%s,omitempty", f.fieldName.LowerCamelComputed)
+
+	// Terraform paths, e.g. { "lifecycle_rule", "*", "transition", "*", "days" } for https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#lifecycle_rule
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: The values of the arrays are differentiating in different points. The biggest difference is on the CnPaths array. Calculating the values from an exact source makes sense however, for this case, it seems that the lists have unique values.

File: pkg/types/builder.go [hunk 5]
```diff
@@ -80,142 +94,70 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
-			tfTag = "-"
-			switch fieldType.String() {
-			case "string", "*string":
-				fieldType = typeSecretKeySelector
-			case "[]string", "[]*string":
-				fieldType = types.NewSlice(typeSecretKeySelector)
-			case "map[string]string", "map[string]*string":
-				fieldType = types.NewMap(types.Universe.Lookup("string").Type(), typeSecretKeySelector)
+		f.fieldNameCamel = f.fieldName.Camel
+		if f.sch.Sensitive {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: NewSensitiveField and NewReferenceField functions were added. Also the AddToResource was added.

As we talked offline, the main goal is the separating the building/constructing *Field object part and adding the constructed *Field to *resource instance. In the new version of refactoring, I tried to address this idea.

File: pkg/types/builder.go [hunk 6]
```diff
@@ -80,142 +94,70 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
 			//  doesn't match anything, it's no-op in late-init logic anyway.
-			if f == fieldPath(tfPaths) {
-				cfg.LateInitializer.AddIgnoredCanonicalFields(fieldPath(cnPaths))
+			if ignoreField == fieldPath(f.tfPaths) {
+				cfg.LateInitializer.AddIgnoredCanonicalFields(fieldPath(f.cnPaths))
 			}
 		}
 
-		fieldType, err := g.buildSchema(sch, cfg, tfPaths, xpPaths, append(names, fieldName.Camel))
+		fieldType, err := g.buildSchema(f.sch, cfg, f.tfPaths, f.xpPaths, append(names, f.fieldName.Camel), f, r)
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Now, we are passing only the f and the function chooses the related values.

Also, append operation was moved to inside of function. By this way, we are doing the append operations together.

File: pkg/types/builder.go [hunk 7]
```diff
@@ -80,142 +94,70 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
-			tfTag = *comment.TerrajetOptions.FieldTFTag
-		}
-		if comment.TerrajetOptions.FieldJSONTag != nil {
-			jsonTag = *comment.TerrajetOptions.FieldJSONTag
-		}
 
+		f.setTags()
+
+		field := types.NewField(token.NoPos, g.Package, f.fieldNameCamel, f.fieldType, false)
 		// NOTE(muvaf): If a field is not optional but computed, then it's
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Moved!

File: pkg/types/builder.go [hunk 8]
```diff
@@ -225,16 +167,131 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
+	}
+	f.jsonTag = name.NewFromCamel(f.fieldNameCamel).LowerCamelComputed
+	if f.sch.Optional {
+		f.fieldType = types.NewPointer(f.fieldType)
+		f.jsonTag += ",omitempty"
+	}
+	return false, nil
+}
+
+func (f *field) setTags() {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: This function was inlined to AddToResource that is responsible adding the constructed field to resource instance.

**Review by muvaf — COMMENTED** [review]
Could you update the testing section of the PR to include a test of a full-set CRD provider, like preview version of jet-aws?

File: pkg/types/builder.go [hunk 9]
```diff
@@ -288,16 +408,24 @@ func (g *Builder) buildSchema(sch *schema.Schema, cfg *config.Resource, tfPath [
 			// If it's optional but also computed, then it means the field has a server
 			// side default but user can change it, so it needs to go to parameters.
 			switch {
-			case isObservation(sch):
+			case isObservation(f.Sch):
 				if obsType == nil {
 					return nil, errors.Errorf("element type of %s is computed but the underlying schema does not return observation type", fieldPath(names))
 				}
 				elemType = obsType
+				if paramType.Underlying().String() != emptyStruct {
```
Comment (muvaf) [inline pkg/types/builder.go:416]: I didn't realize this is fixing https://github.com/crossplane/terrajet/issues/27 but I understood it after syncing with @sergenyalcin . Well done for such clean solution! We probably need to drop a comment here though to make it more clear why we're doing this check.

File: pkg/types/builder.go [hunk 10]
```diff
@@ -62,6 +64,117 @@ type Builder struct {
 	comments twtypes.Comments
 }
 
+// Field represents a field that is built from the Terraform schema.
+// It contains the go field related information such as tags, field type, comment.
+type Field struct {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Moved!

File: pkg/types/builder.go [hunk 11]
```diff
@@ -62,6 +64,117 @@ type Builder struct {
...
+	f.Reference = ref
+
+	f.Comment.Reference = *ref
+	f.Sch.Optional = true
+
+	return f, nil
+}
+
+// Namer represents the parameter and observation name of the resource.
+type Namer struct {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Changed to TypeNames

File: pkg/types/builder.go [hunk 12]
```diff
@@ -80,165 +193,171 @@ func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPa
...
+	}
+	f.JSONTag = name.NewFromCamel(f.FieldNameCamel).LowerCamelComputed
+	if f.Sch.Optional {
+		f.FieldType = types.NewPointer(f.FieldType)
+		f.JSONTag += ",omitempty"
+	}
+	return false, nil
+}
+
+func (f *Field) handleIgnoredFields(cfg *config.Resource) {
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Inlined!

File: pkg/types/builder.go [hunk 13]
```diff
@@ -62,6 +64,117 @@ type Builder struct {
 	comments twtypes.Comments
 }
 
+// Field represents a field that is built from the Terraform schema.
+// It contains the go field related information such as tags, field type, comment.
+type Field struct {
+	Sch                            *schema.Schema
+	FieldName                      name.Name
```
[the remark this answers is not in the document]
Comment (sergenyalcin) [reply pkg/types/builder.go]: Done!

**Pull request merged**